### DISCUSSION TITLE
依赖刷新，service在组件挂载时会被调用两次

### DISCUSSION THREAD

**Discussion by mrdulin** [alibaba/hooks#2425] [Q&A]
尝试使用`useDeepCompareEffect` + `refresh`实现依赖刷新功能，文档： https://ahooks.js.org/zh-CN/hooks/use-request/refresy-deps

```tsx
import React, { useState, useEffect } from "react";
import { useDeepCompareEffect, useRequest } from "ahooks";

const userSchool = (id: string) => {
  switch (id) {
    case "1":
      return "Tsinghua University";
    case "2":
      return "Beijing University";
    case "3":
      return "Zhejiang University";
    default:
      return "";
  }
};

async function getUserSchool(userId): Promise<string> {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve(userSchool(userId));
    }, 1000);
  });
}

const Example1 = () => {
  const [filter] = useState({ userId: "1" });
  const { refresh } = useRequest(() => {
    console.log("[example 1] filter: ", filter);
    return getUserSchool(filter.userId);
  });

  useDeepCompareEffect(() => {
    refresh();
  }, [filter]);

  return <div>example 1</div>;
};

const Example2 = () => {
  const [userId] = useState("1");
  const { refresh } = useRequest(() => {
    console.log("[example 2] userId: ", userId);
    return getUserSchool(userId);
  });

  useEffect(() => {
    refresh();
  }, [userId]);

  return <div>example 2</div>;
};

export default () => {
  return (
    <div>
      <Example1 />
      <Example2 />
    </div>
  );
};
```

组件挂载时，logs:

![image](https://github.com/alibaba/hooks/assets/17866683/baf89a7b-7f27-4866-8fac-8db3ea822c73)

可以看到`<Example1/>`中的service被调用了两次

[codesandbox](https://codesandbox.io/p/sandbox/objective-cherry-2y624f?file=%2FApp.tsx%3A9%2C35&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clqm5jbqo00063j6id637e33u%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clqm5jbqn00023j6igxlmpgv2%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clqm5jbqn00033j6il2jt9fpf%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clqm5jbqn00053j6i0c9sarow%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clqm5jbqn00023j6igxlmpgv2%2522%253A%257B%2522id%2522%253A%2522clqm5jbqn00023j6igxlmpgv2%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clqm5kf7900683j6ikttjcflm%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A9%252C%2522startColumn%2522%253A35%252C%2522endLineNumber%2522%253A9%252C%2522endColumn%2522%253A35%257D%255D%252C%2522filepath%2522%253A%2522%252FApp.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clqm5kf7900683j6ikttjcflm%2522%257D%252C%2522clqm5jbqn00053j6i0c9sarow%2522%253A%257B%2522id%2522%253A%2522clqm5jbqn00053j6i0c9sarow%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clqm5jbqn00043j6ihmo4kxxd%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clqm5jbqn00043j6ihmo4kxxd%2522%257D%252C%2522clqm5jbqn00033j6il2jt9fpf%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clqm5jbqn00033j6il2jt9fpf%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

**Comment by liuyib** [comment]
useRequest 挂载就会请求，然后获取数据后内部状态更新，render 一次。你 useDeepCompareEffect 里又 refresh 了一次请求，再更新一次，这是预期表现。

想要初次只更新一次，把：

```tsx
useDeepCompareEffect(() => {
    refresh();
  }, [filter]);
```

换成：

```tsx
useUpdateEffect(() => {
    refresh();
  }, [filter.userId]);
```

**Reply by mrdulin** [reply]
更新了问题中的代码用于更好的演示

为什么文档中使用`useEffect` + `refresh` +基本数据类型的例子只会调用一次service? 
而，`useDeepCompareEffect` + `refresh` + 引用数据类型会调用两次service

在没看源码的情况下，我武断的认为`useDeepCompareEffect`和`useEffect`唯一的区别就是对deps比较算法，前者使用`lodash.isEqual()`，后者使用`Object.is()`

使用`useDeepCompareEffect`而不是`useUpdateEffect`是因为deps是个嵌套的引用数据类型，是一个对象，具有大量属性，且这些属性的值也是引用数据类型，同时也为了减少代码量，不用将每个属性都罗列在deps中，因此需要使用deepEqual对deps进行比较。

**Reply by liuyib** [reply]
> 为什么文档中使用`useEffect` + `refresh` +基本数据类型的例子只会调用一次service?

你是指这块吗：
![image](https://github.com/alibaba/hooks/assets/38221479/9d469f38-e3aa-4a4e-83de-247520171aaa)

emmmm... 你这么一说，我感觉文档这里写的有问题

---

> 在没看源码的情况下，我武断的认为useDeepCompareEffect和useEffect唯一的区别就是对deps比较算法，前者使用lodash.isEqual()，后者使用Object.is()

这是对的

但 `useDeepCompareEffect` 和 `useUpdateEffect` 的区别就不仅仅是深浅比较的问题了。`useUpdateEffect` mount 时不触发。

感觉文档应该改一下，改成：
![image](https://github.com/alibaba/hooks/assets/38221479/bc08da9e-0ac0-4e04-8825-3f5df839dd0d)

我会再确认下。

> 同时也为了减少代码量，不用将每个属性都罗列在deps中，因此需要使用deepEqual对deps进行比较

这一点目前满足不了哦，应该有个 `useUpdateDeepCompareEffect` mount 时不触发，然后支持深比较。目前的 `useDeepCompareEffect` mount 时仍会触发一次。

**Reply by liuyib** [reply]
@crazylxr  @li-jia-nan  @hchlq @miracles1919 看下呢，我感觉文档是错的，需要改下

**Reply by mrdulin** [reply]
@liuyib OK，我也想是否有个`useDeepCompareUpdateEffect ` hook， 我知道`useUpdateEffect`在componentDidMount时不触发，示例中比较的是`useEffect`和`useDeepCompareEffect`.  `useEffect`在mount和update时会触发，那么文档中的示例`getUserSchool`为何只被调用了一次?
